### ISSUE TITLE
Shell extensions?

### ISSUE THREAD

**Issue by duckcantcode** [retep998/winapi-rs#1041]
This is my first time working w/ the Windows API so bear with me.

Is there a way to make a shell extension in winapi? I want to use [IPropertyStore::SetValue](https://learn.microsoft.com/en-us/windows/win32/api/propsys/nf-propsys-ipropertystore-setvalue?source=recommendations) to add info to the Details pane when you try to view a certain type of file's properties.

It looks like winapi isn't being maintained very much anymore but I thought I would ask.

**Comment by sollyucko** [comment]
What exactly are you asking? That function does exist: https://docs.rs/winapi/latest/winapi/um/propsys/struct.IPropertyStore.html#method.SetValue

**Comment by duckcantcode** [comment]
> What exactly are you asking? That function does exist: https://docs.rs/winapi/latest/winapi/um/propsys/struct.IPropertyStore.html#method.SetValue

@sollyucko Thank you for the quick response :) I'm asking how to make a shell extension that will register this function and allow me to call it. I think I have to make a dll and then register it. What I want to know is how to make that dll in Rust. Do I need to write a `DllMain`? Where do I go from there?

**Comment by duckcantcode** [comment]
I will try to explain better. I want to create a shell extension, with just Rust, that will let me modify how Windows Explorer deals with reading details from a certain file type. But the issue is I don't know how to make something like that.

**Comment by sollyucko** [comment]
You can use `crate-type=cdylib` to generate a C-compatible DLL: https://doc.rust-lang.org/reference/linkage.html

**Comment by duckcantcode** [comment]
Awesome :) now what functions do I add here to make it work with the shell? `DllMain` perhaps? And how do I implement the interface

**Comment by sollyucko** [comment]
Yeah, I guess so. I'm not very familiar with this, but I think these resources should help you get started:

- Well-commented example C++ source code: https://www.tecgraf.puc-rio.br/iup/examples/shell_extensions/ (start with dllmain.cpp)
- "Initializing Shell Extension Handlers": https://learn.microsoft.com/en-us/windows/win32/shell/int-shell-exts

**Comment by duckcantcode** [comment]
This is good. Thank you. I'll let you know how it goes

**Comment by duckcantcode** [comment]
@sollyucko How to implement IClassFactory? https://www.tecgraf.puc-rio.br/iup/examples/shell_extensions/ClassFactory.cpp They did it here. I have to implement https://docs.rs/winapi/latest/winapi/um/unknwnbase/struct.IClassFactory.html somehow

**Comment by sollyucko** [comment]
I'm not sure. Maybe use [`RIDL!`](https://docs.rs/winapi/latest/winapi/macro.RIDL.html) somehow?

**Comment by duckcantcode** [comment]
I'm gonna use https://raw.githubusercontent.com/GabrielMajeri/com-impl-rs for now. So I have to Rewrite-it-in-Rust for everything in that example you sent?

**Comment by duckcantcode** [comment]
I've started to remake all of it. I'm not sure it'll work because the macro doesn't tell me whether I'm using the correct function signatures or not. I may miss a `*mut` or a `*` or a `**` somewhere.

**Comment by duckcantcode** [comment]
I will post my progress at https://github.com/duckfromdiscord/midi-windows-ext periodically.

for `RegCreateKeyExW`, if `phkResult` [is an out](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regcreatekeyexw), why does it ask me to initialize it?

```rust
error[E0381]: used binding `hKey` isn't initialized
  --> src\reg.rs:28:47
   |
20 |     let mut hKey: *mut HKEY;
   |         -------- binding declared here but left uninitialized
...
28 |          KEY_WRITE, &mut security_attributes, hKey, 0 as *mut u32).try_into().unwrap());
   |                                               ^^^^ `hKey` used here but it isn't initialized
   |
help: consider assigning a value
   |
20 |     let mut hKey: *mut HKEY = todo!();
   |    
```

**Comment by sollyucko** [comment]
> So I have to Rewrite-it-in-Rust for everything in that example you sent?

Yeah, probably, unless you want to deal with C++ FFI.

> I may miss a `*mut` or a `*` or a `**` somewhere.

Just keep in mind that `P` or `LP` means `*mut` and `PC` or `LPC` means `*const`, and you should be fine, I think.

> for `RegCreateKeyExW`, if `phkResult` [is an out](https://learn.microsoft.com/en-us/windows/win32/api/winreg/nf-winreg-regcreatekeyexw), why does it ask me to initialize it?

You're trying to directly read the value of an uninitialized variable. Assuming you're trying to have the handle end up directly in `hKey`, not in what in points to, do one of the following:
- Declare `hKey` as `HKEY` and initialize it to some value, e.g. [`std::ptr::null_mut()`](https://doc.rust-lang.org/std/ptr/fn.null_mut.html), and call `RegCreateKeyExW` with `&mut hKey`
- Declare `hKey` as [`MaybeUninit<HKEY>`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html) and initialize it as [`MaybeUninit::uninit()`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.uninit), call `RegCreateKeyExW` with [`hKey.as_mut_ptr()`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.as_mut_ptr), and, if you're completely sure it's initialized, get the `HKEY` out using [`hKey.assume_init()`](https://doc.rust-lang.org/stable/std/mem/union.MaybeUninit.html#method.assume_init-1)

**Comment by duckcantcode** [comment]
Thank you very much.

**Comment by duckcantcode** [comment]
We don't have `StringCchPrintf`

**Comment by sollyucko** [comment]
winapi does have `w(v)sprintf(A/W)`; can you use that instead? The buffer has a fixed length of 1024 bytes though...

[`libc::snprintf`](https://docs.rs/libc/latest/libc/fn.snprintf.html) is probably a much better option.

`format!`/`write!` is another option, but that has different format string syntax and output semantics.

I think the `String...` functions are only defined in `strsafe.h`, not in any built-in object files, but you could write a bit of C code to generate an object file with them, then write custom bindings to link it to your Rust code.

**Comment by duckcantcode** [comment]
> Yeah, I guess so. I'm not very familiar with this, but I think these resources should help you get started:
> 
>     * Well-commented example C++ source code: https://www.tecgraf.puc-rio.br/iup/examples/shell_extensions/ (start with dllmain.cpp)
> 
>     * "Initializing Shell Extension Handlers": https://learn.microsoft.com/en-us/windows/win32/shell/int-shell-exts

Looks like somehow, between the last time I was looking at this code 3 weeks ago, and today, they privated all the code.

**Comment by nichoj10** [comment]
Did this ever work out for you? @duckfromdiscord 

I'm trying to build a Shell extension in Rust using this documentation: https://github.com/MicrosoftDocs/windows-dev-docs/blob/docs/hub/apps/desktop/modernize/integrate-packaged-app-with-file-explorer.md#implement-the-shell-extension.

Would love any pointers if you have them

**Comment by duckcantcode** [comment]
It did not. Let me know if you find anything.

**Comment by sollyucko** [comment]
> > Yeah, I guess so. I'm not very familiar with this, but I think these resources should help you get started:
> > ```
> > * Well-commented example C++ source code: https://www.tecgraf.puc-rio.br/iup/examples/shell_extensions/ (start with dllmain.cpp)
> > 
> > * "Initializing Shell Extension Handlers": https://learn.microsoft.com/en-us/windows/win32/shell/int-shell-exts
> > ```
> 
> Looks like somehow, between the last time I was looking at this code 3 weeks ago, and today, they privated all the code.

What do you mean by this, and is this still an issue for you? I can still see code at both links, e.g. https://learn.microsoft.com/en-us/windows/win32/shell/int-shell-exts#implementing-ishellextinit, https://www.tecgraf.puc-rio.br/iup/examples/shell_extensions/dllmain.cpp

**Comment by duckcantcode** [comment]
I guess the link is back now, it was down when I tried it 🤔
